File: app/src/main/java/com/stevenfrew/beatprompter/SongLoaderTask.java
package com.stevenfrew.beatprompter;

import android.os.Handler;

import com.stevenfrew.beatprompter.event.CancelEvent;

import java.io.IOException;

class SongLoaderTask extends Task {
    private CancelEvent mCancelEvent=null;
    private SongLoadInfo mSongLoadInfo=null;
    private static Song mCurrentSong=null;
    private Handler mSongLoadHandler=null;
    private boolean mRegistered;

    private final static Object mCurrentSongSync=new Object();
    private final Object mSongLoadHandlerSync = new Object();
    private final Object mLoadingSongFileSync = new Object();
    private final Object mCancelEventSync = new Object();

    SongLoaderTask()
    {
        super(true);
    }

    private Handler getSongLoadHandler()
    {
        synchronized (mSongLoadHandlerSync)
        {
            return mSongLoadHandler;
        }
    }
    private void setSongLoadHandler(Handler handler)
    {
        synchronized (mSongLoadHandlerSync)
        {
            mSongLoadHandler=handler;
        }
    }


    static Song getCurrentSong()
    {
        synchronized (mCurrentSongSync)
        {
            return mCurrentSong;
        }
    }

    static void setCurrentSong(Song song)
    {
        synchronized (mCurrentSongSync)
        {
            mCurrentSong=song;
            System.gc();
        }
    }

    private SongLoadInfo getLoadingSongFile()
    {
        synchronized (mLoadingSongFileSync)
        {
            SongLoadInfo result=mSongLoadInfo;
            mSongLoadInfo=null;
            return result;
        }
    }
    private CancelEvent getCancelEvent()
    {
        synchronized (mCancelEventSync)
        {
            return mCancelEvent;
        }
    }
    private void setCancelEvent(CancelEvent cancelEvent)
    {
        synchronized (mCancelEventSync)
        {
            mCancelEvent=cancelEvent;
        }
    }

    public void doWork()
    {
        SongLoadInfo sli=getLoadingSongFile();
        if(sli!=null) {
            System.gc();
            Handler songLoadHandler=getSongLoadHandler();
            CancelEvent cancelEvent = getCancelEvent();
            try {
                SongLoader loader=new SongLoader(sli,cancelEvent,songLoadHandler,mRegistered);
                Song loadedSong = loader.load();
                if (cancelEvent.isCancelled())
                    songLoadHandler.obtainMessage(EventHandler.SONG_LOAD_CANCELLED).sendToTarget();
                else {
                    setCurrentSong(loadedSong);
                    songLoadHandler.obtainMessage(EventHandler.SONG_LOAD_COMPLETED).sendToTarget();
                }
            }
            catch(IOException ioe)
            {
                songLoadHandler.obtainMessage(EventHandler.SONG_LOAD_FAILED,ioe.getMessage()).sendToTarget();
            }
            System.gc();
        }
        else
            try {
                // Nothing to do, wait a bit
                Thread.sleep(250);
            }
            catch(InterruptedException ignored)
            {
            }
    }
    void loadSong(SongLoadInfo sli, Handler handler, CancelEvent cancelEvent,boolean registered)
    {
        setSongLoadHandler(handler);
        synchronized (mLoadingSongFileSync)
        {
            CancelEvent existingCancelEvent=getCancelEvent();
            if(existingCancelEvent!=null)
                existingCancelEvent.set();
            setCancelEvent(cancelEvent);
            mRegistered=registered;
            mSongLoadInfo=sli;
        }
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/song/load/SongLoaderTask.kt
package com.stevenfrew.beatprompter.song.load

import android.util.Log
import com.stevenfrew.beatprompter.BeatPrompterApplication
import com.stevenfrew.beatprompter.EventHandler
import com.stevenfrew.beatprompter.song.Song
import com.stevenfrew.beatprompter.Task
import com.stevenfrew.beatprompter.cache.parse.SongParser

class SongLoaderTask : Task(true) {
    private val mJobSync = Any()
    private var mLoadingJob: SongLoadJob? = null
    private var mCuedJob: SongLoadJob? = null

    override fun doWork() {
        val slj:SongLoadJob?=
        {
            synchronized(mJobSync) {
                mLoadingJob=mCuedJob
                mCuedJob=null
                mLoadingJob
            }
        }.invoke()
        if (slj != null) {
            System.gc()
            try {
                Log.d(BeatPrompterApplication.TAG,"Starting to load '${slj.mSongLoadInfo.mSongFile.mTitle}'.")
                val loadedSong = SongParser(slj.mSongLoadInfo,slj.mCancelEvent,slj.mHandler,slj.mRegistered).parse()
                if (slj.mCancelEvent.isCancelled)
                    throw SongLoadCancelledException()
                Log.d(BeatPrompterApplication.TAG,"Song was loaded successfully.")
                currentSong = loadedSong
                slj.mHandler.obtainMessage(EventHandler.SONG_LOAD_COMPLETED,slj.mSongLoadInfo.mLoadID).sendToTarget()
            } catch (e: SongLoadCancelledException) {
                Log.d(BeatPrompterApplication.TAG,"Song load was cancelled.")
                slj.mHandler.obtainMessage(EventHandler.SONG_LOAD_CANCELLED).sendToTarget()
            } catch (e: Exception) {
                Log.d(BeatPrompterApplication.TAG,"Song load failed.")
                slj.mHandler.obtainMessage(EventHandler.SONG_LOAD_FAILED, e.message).sendToTarget()
            }
            finally {
                synchronized(mJobSync) {
                    mLoadingJob=null
                }
            }

            System.gc()
        } else
            try {
                // Nothing to do, wait a bit
                Thread.sleep(250)
            } catch (ignored: InterruptedException) { }
    }

    fun loadSong(songLoadJob:SongLoadJob) {
        synchronized(mJobSync) {
            Log.d(BeatPrompterApplication.TAG,"In SongLoaderTask.loadSong() ...")
            if(mLoadingJob!=null) {
                Log.d(BeatPrompterApplication.TAG,"Setting cancel event for load of '${mLoadingJob?.mSongLoadInfo?.mSongFile?.mTitle}'")
                mLoadingJob?.mCancelEvent?.set()
            }
            mCuedJob=songLoadJob
        }
    }

    companion object {
        private var mCurrentSong: Song? = null

        private val mCurrentSongSync = Any()

        var currentSong: Song?
            get() = synchronized(mCurrentSongSync) {
                return mCurrentSong
            }
            private set(song) = synchronized(mCurrentSongSync) {
                mCurrentSong = song
                System.gc()
            }
    }
}
